metrics

melekes opened the pull request:
Refs #986 

![screen shot 2018-06-15 at 15 20 43](https://user-images.githubusercontent.com/1282182/41465654-12c9ad7a-70b0-11e8-8e9c-9b16e70d6bef.png)

melekes commented on docs/architecture/adr-010-monitoring.md:
```diff
@@ -0,0 +1,103 @@
+future. It's API looks different from go-kit and Prometheus, but looks like it
+covers everything we need.
+
+Unfortunately, OpenCensus go client does not define any
```
looks like it's too much work. so maybe we should go with {go-kit + prometheus}

xla commented:
All in favor of an initial setup with gokit and prometheus. The two concepts are not interchangeable. Metrics, together with logs and tracing make up the larger story of observability. Where logs and metrics have the biggest return for the effort to make. Tracing is very powerful but puts a lot of pressure on the supporting infrastructure where a metrics endpoint in our binaries gives instant value.

xla commented on consensus/state.go:
```diff
@@ -388,6 +392,7 @@ func (cs *ConsensusState) SetProposalAndBlock(proposal *types.Proposal, block *t
 
 func (cs *ConsensusState) updateHeight(height int64) {
 	cs.Height = height
+	cs.metrics.Height.Add(float64(height - cs.Height))
```
Same as with logging, we keep interleaving non-business concerns with our business logic, which makes for harder to maintain code. I know this would require a bigger refactor but this might be a good time to think about better composability. Which could be done with the more rigirous use of interfaces and then layered decoratored implementations for the different concerns (logic, logging, metrics). This [video](https://www.youtube.com/watch?v=xyDkyFjzFVc) gives an intro how that looks and I can give some examples how it is done in production code.

melekes replied:
Video is great. I love it.

xla commented on docs/architecture/adr-010-monitoring.md:
```diff
@@ -0,0 +1,103 @@
+|   | Name                                    | Type    |                                                                               |
+| - | --------------------------------------- | ------- | ----------------------------------------------------------------------------- |
+| A | height                                  | Counter |                                                                               |
+| A | validators:<height>                     | Gauge   | Number of validators who signed                                               |
```
`<height>` has a couple of occurrences, which I suspect will lead to high cardinality in our metrics, something which can be a problem for long-running processes. Are we going to cap how many of them we keep around, or do we just expose the latest height?

melekes replied:
I thought of it as a label. Like `validators` with a label `height-11`. Sorry for confusion. What do you think it?

xla commented:
Follow-up to https://github.com/tendermint/tendermint/pull/1737#discussion_r195034746: Generally any cardinality beyond 10s of things is going to put a lot of pressure on the collectors. See the [label section](https://prometheus.io/docs/practices/naming/#labels) in the prometheus documentation. I was under the impression that `height` is one of those unbounded set of values.

melekes commented on docs/architecture/adr-011-monitoring.md:
```diff
@@ -0,0 +1,105 @@
+
+|   | Name                                    | Type    | Labels                    | Description                                                                   |
+| - | --------------------------------------- | ------- | ------------------------- | ----------------------------------------------------------------------------- |
+| A | height                                  | Counter | height-X                  |                                                                               |
```
list needs to be updated

melekes commented on node/node.go:
```diff
@@ -81,10 +85,97 @@ func DefaultNewNode(config *cfg.Config, logger log.Logger) (*Node, error) {
 	)
 }
 
+// MetricsProvider returns a consensus and p2p Metrics.
```
move to a separate package?

melekes commented on types/block.go:
```diff
@@ -135,6 +135,15 @@ func (b *Block) HashesTo(hash []byte) bool {
 
+// Size returns size of the block in bytes.
+func (b *Block) Size() int {
+	bz, err := cdc.MarshalBinaryBare(b)
```
expensive to call for each block?

xla replied:
Can we make a call on this? Do we know about the cost?

melekes replied:
![screen shot 2018-06-20 at 11 16 05](https://user-images.githubusercontent.com/1282182/41643102-615b08d4-747b-11e8-8ab8-122fb44143fa.png)

~10ms

melekes commented:
Should we add a config option?

xla commented on node/node.go:
```diff
@@ -81,10 +85,102 @@ func DefaultNewNode(config *cfg.Config, logger log.Logger) (*Node, error) {
+
+// DefaultMetricsProvider returns consensus, p2p and mempool Metrics build
+// using Prometheus client library.
+func DefaultMetricsProvider() (*cs.Metrics, *p2p.Metrics, *mempl.Metrics) {
```
This seems to be the wrong place for setting those up, as the metrics are observed in the packages themselves they should live there. Also this will grow into a hug construct with more metrics/packages, it would live better in the `metrics.go` int he respective package with the sane parametrisation.

melekes replied:
1. I don't want to bring prometheus package inside our internal packages (cs, mempool, ..)
2. The reason we have *Provider is because some people use Tendermint as a library and for them to use different DB or metrics, it's much easier to only change 1 function - NewNode instead of changing all source code of Tendermint

xla replied:
That's very reasonable. I still don't believe we need this huge constructor for metrics, check my comment on the state storage above.

melekes replied:
moved it to a separate package (`metrics/prometheus`)

xla commented on node/node.go:
```diff
@@ -122,6 +218,7 @@ func NewNode(config *cfg.Config,
 	clientCreator proxy.ClientCreator,
 	genesisDocProvider GenesisDocProvider,
 	dbProvider DBProvider,
+	metricsProvider MetricsProvider,
```
`NewNode` already takes the config, no need to extend its surface even further we can do the setup of what metric to use for what component down below.

melekes commented:
@xla thanks for the feedback! greatly appreciated

> the metrics endpoitn should not be on the same port as RPC

agree, changed.

> we should avoid extending the surface of our components. 

you're right. check out my last commits where I changed constructors to accept the list of options aka https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

melekes commented on consensus/state.go:
```diff
@@ -1291,6 +1300,43 @@ func (cs *ConsensusState) finalizeCommit(height int64) {
 	// * cs.StartTime is set to when we will start round0.
 }
 
+func (cs *ConsensusState) recordMetrics(height int64, block *types.Block) {
```
But where do we store the state? In the reactor?

xla replied:
Ideally we don't store the state anywhere instead keep it enclosed in the outter scope. To illustrate:

``` go
type recordMetricsFunc func(int64, *types.Block, *types.ValidatorSet)

func promMetrics() recordMetricsFunc {
	metrics := &Metrics{}

	return func(int64, *types.Block, *types.ValidatorSet) {
		// update metrics
	}
}

func nopMetrics() recordMetricsFunc {
	return func(int64, *types.Block, *types.ValidatorSet) {
		// nop
	}
}
```

melekes replied:
that's really smart 👍 let's see what I can do

melekes replied:
the problem is not all the metrics can be recorded in this one function. some of the like round increments do not belong to this function... but the idea itself is beautiful.

xla replied:
There are couple of ways to go about it, one is to introduce specialised functions and another have the `recordMetrics` take all possible arguments and decide which ones to record based on what has been passed in, e.g. `if block == nil { return }`.

xla commented on node/node.go:
```diff
@@ -527,6 +650,22 @@ func (n *Node) startRPC() ([]net.Listener, error) {
 
+// StartPrometheusServer starts a Prometheus HTTP server, listening for metrics
+// collectors on addr.
+func (n *Node) StartPrometheusServer(addr string) *http.Server {
```
Does this need to be public?

melekes replied:
no

xla commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,102 @@
+package prometheus
```
I know we already had a discussion and on second thought I don't agree with the sentiment that this should not go in the respective packages. The declaration of the structs, the default nop implementation and the general ownership of the metrics are already there. We import prometheus anyway in Tendermint and as long as we don't have pressurign use-cases for single packages which require prometheus not to be imported I think we should not make decisions based on it. Even if we decide we want "cleaner" imports we can always make sub-packages with the metric providers. As it stands right now it feels dislocated and is harder to maintain as there are multiple places to keep track, which is unnecessary burden.

xla commented on consensus/state.go:
```diff
@@ -1291,6 +1310,43 @@ func (cs *ConsensusState) finalizeCommit(height int64) {
+	if height > 1 {
+		lastBlockMeta := cs.blockStore.LoadBlockMeta(height - 1)
+		cs.metrics.BlockIntervalSeconds.
+			Observe(block.Time.Sub(lastBlockMeta.Header.Time).Seconds())
```
This is an unconventional way of splitting long lines. Preferrably:

``` go
	cs.metrics.BlockIntervalSeconds.Observe(
		block.Time.Sub(lastBlockMeta.Header.Time).Seconds(),
	)
```

xla commented on consensus/state.go:
```diff
@@ -115,10 +115,13 @@ type ConsensusState struct {
 
 // NewConsensusState returns a new ConsensusState.
-func NewConsensusState(config *cfg.ConsensusConfig, state sm.State, blockExec *sm.BlockExecutor, blockStore sm.BlockStore, mempool sm.Mempool, evpool sm.EvidencePool) *ConsensusState {
+func NewConsensusState(config *cfg.ConsensusConfig, state sm.State, blockExec *sm.BlockExecutor, blockStore sm.BlockStore, mempool sm.Mempool, evpool sm.EvidencePool, options ...func(*ConsensusState)) *ConsensusState {
```
This is growing quite large can we split it in multiple lines?

xla commented on node/node.go:
```diff
@@ -208,11 +229,24 @@ func NewNode(config *cfg.Config,
-	mempool := mempl.NewMempool(config.Mempool, proxyApp.Mempool(), state.LastBlockHeight)
-	mempool.InitWAL() // no need to have the mempool wal during tests
+	mempool := mempl.NewMempool(config.Mempool, proxyApp.Mempool(), state.LastBlockHeight,
+		mempl.WithMetrics(memplMetrics))
```
Better to split like:

``` go
mempool := mempl.NewMempool(
    config.Mempool,
    proxyApp.Mempool(),
    state.LastBlockHeight,
    mempl.WithMetrics(memlMetrics),
)
```

melekes commented:
@xla thanks again for a thorough review. next time feel free to make changes yourself instead of spending time writing comments.

> re https://github.com/tendermint/tendermint/pull/1737#discussion_r196617652

if you're sure about this, please go ahead and implement
